### DISCUSSION TITLE
error when using update_and_rebuild_libmesh.sh using msys2 on windows

### DISCUSSION THREAD

**Discussion by starkekr** [libMesh/libmesh#3471] [Q&A]
I am following the MOOSE installation instructions [here](https://mooseframework.inl.gov/getting_started/installation/msys2.html). I have progressed through the installation instructions and am stuck on the libmesh installation. I am following the instructions for building libmesh as a submodule of MOOSE.

I have the following errors when I run ./scripts/update_and_rebuild_libmesh.sh at the ending step of the rebuilding MOOSE.

make[2]: Leaving directory '/home/stark/projects/moose/libmesh/build/contrib/eigen/eigen'
make[1]: *** [Makefile:1033: all-recursive] Error 1
make[1]: Leaving directory '/home/stark/projects/moose/libmesh/build/contrib'
make: *** [Makefile:33131: all-recursive] Error 1

It gets caught up making eigen. I would greatly appreciate any help.

**Comment by starkekr** [comment]
I have disabled eigen and netcdf using --disable-eigen --disable-netcdf but now I have a new problem. 

../src/base/libmesh.C: In constructor 'libMesh::LibMeshInit::LibMeshInit(int, const char* const*, TIMPI::communicator, int)':
../src/base/libmesh.C:648:3: error: 'setenv' was not declared in this scope; did you mean 'getenv'?
  648 |   setenv("HDF5_USE_FILE_LOCKING", "FALSE", /*overwrite=false*/0);
      |   ^~~~~~
      |   getenv
make[1]: *** [Makefile:26063: src/base/libmesh_opt_la-libmesh.lo] Error 1


any suggestions?

**Comment by jwpeterson** [comment]
Looks like we picked up this msys2 regression in d457766c25. I guess a temporary fix could be to change the line:
```
#ifdef LIBMESH_HAVE_HDF5
```
to
```
#if defined(LIBMESH_HAVE_HDF5) && !defined(_MSC_VER)
```
That should simply skip the setenv command when you are compiling on windows. Not sure what the correct equivalent to a `setenv` call on Windows would actually be.

**Reply by starkekr** [reply]
I can check to see if this works but for now I just disable hdf5 and then I get the following error

../src/base/print_trace.C:55:13: error: conflicting declaration 'typedef int pid_t'
   55 | typedef int pid_t;
      |             ^~~~~
In file included from C:/msys64/mingw64/include/pthread.h:64,
                 from C:/msys64/mingw64/include/c++/12.2.0/x86_64-w64-mingw32/bits/gthr-default.h:35,
                 from C:/msys64/mingw64/include/c++/12.2.0/x86_64-w64-mingw32/bits/gthr.h:148,
                 from C:/msys64/mingw64/include/c++/12.2.0/ext/atomicity.h:35,
                 from C:/msys64/mingw64/include/c++/12.2.0/bits/ios_base.h:39,
                 from C:/msys64/mingw64/include/c++/12.2.0/ios:42,
                 from C:/msys64/mingw64/include/c++/12.2.0/ostream:38,
                 from C:/msys64/mingw64/include/c++/12.2.0/iostream:39,
                 from ./include/libmesh/print_trace.h:25,
                 from ../src/base/print_trace.C:46:
C:/msys64/mingw64/include/sys/types.h:68:17: note: previous declaration as 'typedef _pid_t pid_t'
   68 | typedef _pid_t  pid_t;


any thoughts on this one?

**Reply by jwpeterson** [reply]
That line was added in b9ec4128a4 by @mkhon, ironically as part of his updates to get libmesh compiling under Windows. It looks like there is a pthread.h on your system, pulled in by the `<iostream>` inclusion, that already defines the `pid_t` type, which causes the error message. I suppose for now you could simply comment out the offending line in print_trace.C and that would fix the problem.

I see that we only actually use the `pid_t` type in one line in this file:
```
      pid_t this_pid = getpid();
```
so, rather than defining any type at all, we should probably just use `auto` on that line? @roystgnr

**Reply by starkekr** [reply]
This same problem appeared in src/mesh/namebased_io.C:56:13 as well. I commented it out as well. This seems to fixed things and I am now able to build libmesh but for Moose to build, it appears that I need eigen as well. On the Moose side, I am getting the following errors

C:/msys64/home/stark/projects/moose/framework/build/header_symlinks/MooseTypes.h:35:10: fatal error: Eigen/Core: No such file or directory
   35 | #include "Eigen/Core"

But I believe this is because I disabled eigen in the libmesh build. When I enable eigen on the libmesh side I get the following error 

Making all in eigen/eigen
make[2]: Entering directory '/home/stark/projects/libmesh/installed/contrib/eigen/eigen'
make[2]: *** No rule to make target 'Eigen/Householder', needed by 'all-am'.  Stop.
make[2]: Leaving directory '/home/stark/projects/libmesh/installed/contrib/eigen/eigen'
make[1]: *** [Makefile:1033: all-recursive] Error 1
make[1]: Leaving directory '/home/stark/projects/libmesh/installed/contrib'
make: *** [Makefile:33131: all-recursive] Error 1

**Reply by jwpeterson** [reply]
I created PR #3473 to fix the `pid_t` typedef issue. 

Regarding the Eigen error, it sounds like Eigen is not optional for MOOSE currently, so you won't be able to build libmesh without Eigen if you want to use MOOSE. I'm not sure what is actually causing the error you are seeing above, since there is nothing to compile for Eigen (it is a header-only library). 

I'm wondering if it is some simple filesystem issue with symbolic links in the git submodule that we use for Eigen... currently we have `contrib/eigen/eigen -> contrib/eigen/gitshim` and `contrib/eigen/gitshim/Eigen -> contrib/eigen/git/Eigen` so if symlinks aren't working for whatever reason, I could see this maybe giving the error you reported.

**Reply by jwpeterson** [reply]
I also opened #3474 to fix the `setenv` error that you encountered.

**Reply by starkekr** [reply]
I tried to verify the symlinks that you specified. I don't know if this is the correct way of doing this but I have attached a picture of my terminal window
![image](https://user-images.githubusercontent.com/118568957/215225000-f6682a86-13ba-47d1-9ba6-c0a1d31838c2.png)

I also tried passing the eigen path to the configure using
 ../configure --with-methods='opt' --prefix=$HOME/projects/libmesh/installed --with-static=yes --with-shared=no --enable-static --disable-shared --enable-unique-id --disable-maintainer-mode --enable-petsc-hypre-required --enable-metaphysicl-required --disable-netcdf --disable-netcdf-4  --with-eigen-include=/home/stark/projects/libmesh/contrib/eigen/3.3.9/Eigen/src && make && make install

but it didn't work. I may be using the wrong directory though.

**Reply by starkekr** [reply]
Ok, so i got past the eigen problem by not relying on the symlinks and just copying the eigen files to /build/contrib/eigen/eigen. Its messy but it works. Now when I went to build MOOSE I get the following error

C:/msys64/home/stark/projects/moose/framework/src/outputs/formatters/ExodusFormatter.C:18:10: fatal error: libmesh/exodusII.h: No such file or directory
   18 | #include "libmesh/exodusII.h"

So I think I need to build libmesh with netcdf enabled or maybe there is some way to disable netcdf on the MOOSE side? The ideal scenario would be that I compile libmesh with netcdf enabled so I started trying to do that. But I ran into another problem, I run the configure script with the following additional arguments:

 --enable-netcdf-required NETCDF_DIR=C:/msys64/mingw64

and the the configure runs fine. Once I run make I get the following error:


make[4]: Entering directory '/home/stark/projects/libmesh/build/contrib/netcdf/v4/libsrc'
  CC       libnetcdf3_la-memio.lo
../../../../../contrib/netcdf/v4/libsrc/memio.c: In function 'memio_new':
../../../../../contrib/netcdf/v4/libsrc/memio.c:148:7: error: unknown type name 'SYSTEM_INFO'
  148 |       SYSTEM_INFO info;
      |       ^~~~~~~~~~~
../../../../../contrib/netcdf/v4/libsrc/memio.c:149:7: warning: implicit declaration of function 'GetSystemInfo' [-Wimplicit-function-declaration]
  149 |       GetSystemInfo (&info);
      |       ^~~~~~~~~~~~~
../../../../../contrib/netcdf/v4/libsrc/memio.c:150:22: error: request for member 'dwPageSize' in something not a structure or union
  150 |       pagesize = info.dwPageSize;

Any thoughts?

**Reply by jwpeterson** [reply]
We don't currently support building with system NetCDF (or any NetCDF other than the one distributed with libmesh). So I don't think this approach of using `NETCDF_DIR=C:/msys64/mingw64` is likely to work unfortunately. Can you remind me what the reason for not using the NetCDF that comes with libmesh is? Is it also a symlink issue?

**Comment by mkhon** [comment]
@jwpeterson my changes were to compile libmesh using Visual Studio. When msys2 is used many posix headers are available, so pid_t definition should be additionally ifdef'ed out for msys2
